src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -174,10 +174,6 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
 
             private void OnDiagnosticsUpdated(DiagnosticsUpdatedArgs e)
             {
-                // First see if this is a document/project removal.  If so, clear out any state we
-                // have associated with any analyzers we have for that document/project.
-                ProcessRemovedDiagnostics(e);
```

Comment:
More evidence that getting this right is extremely difficult :-)

A core scenario for diagnostics is 'let me know the correct diagnostics for this buffer'.  It's completely unclear how to do that with the existing API.   This is unlike practically every other API we have for getting information about a file.  

I think this functionality should be encapsulated into the service itself. It can do all the appropriate logic we're talking about here.